cmake/EthInstallation.cmake

```diff
@@ -0,0 +1,41 @@
+# Install internal libraries.
+#
+# Installation will install the internal libraries, headers AND dependencies to the configured locations
+# After installation, the libraries can be linked against using `find_package(Solidity)`.
+# Note that this feature is expected to work, but unsupported.
+
+# Installing internal libraries
+install(TARGETS evmasm DESTINATION "${CMAKE_INSTALL_LIBDIR}")
+install(TARGETS langutil DESTINATION "${CMAKE_INSTALL_LIBDIR}")
+install(TARGETS smtutil DESTINATION "${CMAKE_INSTALL_LIBDIR}")
+install(TARGETS libsolc DESTINATION "${CMAKE_INSTALL_LIBDIR}")
+install(TARGETS solidity DESTINATION "${CMAKE_INSTALL_LIBDIR}")
+install(TARGETS solutil DESTINATION "${CMAKE_INSTALL_LIBDIR}")
+install(TARGETS yul DESTINATION "${CMAKE_INSTALL_LIBDIR}")
+
+# Install internal headers
+foreach(ETH_COMPONENT IN LISTS ETH_COMPONENTS)
+  install(DIRECTORY ${CMAKE_SOURCE_DIR}/${ETH_COMPONENT}
+          DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+          PATTERN "*.cpp" EXCLUDE
+          PATTERN "CMakeLists.txt" EXCLUDE
+          PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE
+  )
+endforeach()
+
+# install deps headers
+foreach(DEP_DIR concepts json meta range std)
+  install(DIRECTORY ${CMAKE_BINARY_DIR}/deps/include/${DEP_DIR}
+    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+    PATTERN "*.cpp" EXCLUDE
+    PATTERN "*.txt" EXCLUDE
+    PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE
+  )
+endforeach()
+
+install(DIRECTORY ${CMAKE_BINARY_DIR}/_deps/fmtlib-src/include/fmt
+  DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}
+  PATTERN "*.cpp" EXCLUDE
+  PATTERN "*.txt" EXCLUDE
+  PERMISSIONS OWNER_READ OWNER_WRITE OWNER_EXECUTE GROUP_READ GROUP_EXECUTE WORLD_READ WORLD_EXECUTE
```

Comment:
Ok, I see now that I may have misunderstood you a bit in the issue :-).

I don't think we can reasonably install dependency headers here, that's very bad practice even if this is disabled by default. Now, of course, the internal headers do depend on these dependency headers, but those internal headers also have no stability guarantee whatsoever, so we don't want to recommend to anybody to actually use them as interface.

The only interface that we *do* keep stable is the C-interface in https://github.com/ethereum/solidity/blob/develop/libsolc/libsolc.h - that header does not depend on any dependency or any other internal header.

But the exposed interface there is of course quite inconvenient - it's a wrapper around the standard-json interface to the compiler and requires serializing and deserializing compiler input and output to json.

May I ask what's your use case, i.e. which parts of the compiler do you actually use and for what purpose? Given that you expose the internal headers here, I'm assuming that you *don't* just restrict yourself to the libsolc.h interface?

Reply:
Thanks for your PR review and yeah, installing dep headers is indeed a bad practice just to make internal modules work as imported targets. My lab is working on a project which basically compiles Solidity sources into Petri Net, then do validations(eg. concurrency safety analysis) with our own Petri Net test framework. The main use cases involved are as such: 
- Lexing and parsing Solidity sources with CompilerStack.
- Implementing our own ASTVisitor and ASTConstVisitor classes for code emitting, etc...
- Using modules such as FileReader as our base IO components

Reply:
Thanks for your PR review and yeah, installing dep headers is indeed a bad practice just to make internal modules work as imported targets. My lab is working on a project which basically compiles Solidity sources into Petri Net, then do validations(eg. concurrency safety analysis) with our own Petri Net test framework. The main use cases involved are as such:

Lexing and parsing Solidity sources with CompilerStack.
Implementing our own ASTVisitor and ASTConstVisitor classes for code emitting, etc...
Using modules such as FileReader as our base IO components

Reply:
Thanks for your PR review and yeah, installing dep headers is indeed a bad practice just to make internal modules work as imported targets. My lab is working on a project which basically compiles Solidity sources into Petri Net, then do validations(eg. concurrency safety analysis) with our own Petri Net test framework. The main use cases involved are as such: 
- Lexing and parsing Solidity sources with CompilerStack.
- Implementing our own ASTVisitor and ASTConstVisitor classes for code emitting, etc...
- Using modules such as FileReader as our base IO components